sw/host/tests/chip/gpio/src/sleep_pin_wake.rs

```diff
@@ -0,0 +1,184 @@
+    Pad { name: "IOC7",  valid_cw310: false, valid_silicon: false },  // MIO Pad 29
+    Pad { name: "IOC8",  valid_cw310: false, valid_silicon: false },  // MIO Pad 30
+    Pad { name: "IOC9",  valid_cw310: false, valid_silicon: true },   // MIO Pad 31
+    Pad { name: "IOC10", valid_cw310: false, valid_silicon: true },   // MIO Pad 32
+    Pad { name: "IOC11", valid_cw310: false, valid_silicon: true },   // MIO Pad 33
+    Pad { name: "IOC12", valid_cw310: false, valid_silicon: true },   // MIO Pad 34
+    Pad { name: "IOR0",  valid_cw310: false, valid_silicon: false },  // MIO Pad 35
+    Pad { name: "IOR1",  valid_cw310: false, valid_silicon: false },  // MIO Pad 36
+    Pad { name: "IOR2",  valid_cw310: false, valid_silicon: false },  // MIO Pad 37
+    Pad { name: "IOR3",  valid_cw310: false, valid_silicon: false },  // MIO Pad 38
+    Pad { name: "IOR4",  valid_cw310: false, valid_silicon: false },  // MIO Pad 39
+    Pad { name: "IOR5",  valid_cw310: false, valid_silicon: true },   // MIO Pad 40
+    Pad { name: "IOR6",  valid_cw310: true,  valid_silicon: true },   // MIO Pad 41
+    Pad { name: "IOR7",  valid_cw310: true,  valid_silicon: true },   // MIO Pad 42
+    Pad { name: "IOR10", valid_cw310: true,  valid_silicon: true },   // MIO Pad 43
+    Pad { name: "IOR11", valid_cw310: true,  valid_silicon: true },   // MIO Pad 44
+    Pad { name: "IOR12", valid_cw310: true,  valid_silicon: true },   // MIO Pad 45
+    Pad { name: "IOR13", valid_cw310: true,  valid_silicon: true },   // MIO Pad 46
+];
+
+fn is_valid_pad(pad: &Pad, interface: &str) -> bool {
+    match interface {
+        "hyper310" => pad.valid_cw310,
+        "teacup" => pad.valid_silicon,
+        _ => false,
+    }
+}
+
+fn sleep_pin_wake_test(
+    opts: &Opts,
+    transport: &TransportWrapper,
+    sival_mio_pad_addr: u32,
+    sival_wakeup_detector_idx_addr: u32,
+) -> Result<()> {
+    let interface = opts.init.backend_opts.interface.as_str();
+    let uart = transport.uart("console")?;
+    uart.set_flow_control(true)?;
+
+    log::info!("Starting host side");
+
+    let detector_match =
+        UartConsole::wait_for(&*uart, r"detector (\d+) is selected", opts.timeout)?;
+    let detector: usize = detector_match[1].parse()?;
+
+    UartConsole::wait_for(&*uart, "SiVal: waiting for commands", opts.timeout)?;
+
+    let mut rand_pad: usize;
+    loop {
+        rand_pad = rand::thread_rng().gen_range(0..MIO_PADS.len());
+        if is_valid_pad(&MIO_PADS[rand_pad], interface) {
+            break;
+        }
+    }
+
+    MemWriteReq::execute(&*uart, sival_mio_pad_addr, &[rand_pad as u8])?;
+
+    let pad_sel_match =
+        UartConsole::wait_for(&*uart, r"Pad Selection: (\d) / (\d+)\r\n", opts.timeout)?;
+    let pad_sel: usize = pad_sel_match[2].parse()?;
+    let wake_pin = transport.gpio_pin(MIO_PADS[pad_sel].name)?;
+
+    wake_pin.set_mode(PinMode::PushPull)?;
+    wake_pin.write(false)?; // Don't wake up yet.
```

Comment:
you might run into issue if you do this because by the time you configure the pin, the device might have already gone to sleep (and wake up if you are unlucky?). It think it would be better if the device waited until the host tells it that it has configured the pin and it can go to sleep.

Reply:
The printing of the `LOG_INFO("Entering low power mode.");` is going to take a bit of time. I'm not sure this will ever be a problem in practise?

Reply:
When the CI is under heavy loads, we have been seeing all kinds of delays, and this leads to flaky tests. I would prefer if the tests did the right thing.